KodrAus/auto_impl
# Rewrite most of the crate with updated dependencies

So I'm basically done with rewriting everything. Of course I didn't just rewrite it (just) for fun, but to update `syn`, `quote` and `proc_macro`. We now use the following versions (the newest as of time of writing):

- proc-macro2 0.4.6
- quote 0.6.3
- syn 0.14.4

The new implementation can now compile the `compile_test` examples (and more). **I guess I reached feature parity**.

I think the old and new implementations are somewhat different:

- The new implementation uses much less own data structures and instead uses `syn` structures to pass information around. 
- Most functionality in the new version is implemented in free functions which usually receive `syn` data structures and return a token stream. 
- The new implementation uses `proc_macro::TokenStream` instead of the `proc_macro2` version where possible in a useful way.

I managed to reduce total line count despite my code having more comments and adding two examples. I guess this is simply due to more ergonomic APIs by `syn` and `proc_macro`. 

---

As far as I can tell, I improved diagnostics in the new version. Example:

![image](https://user-images.githubusercontent.com/7419664/42412929-226b6a3a-8216-11e8-840b-36a373e3284e.png)

---
## Notes for the Review

I know this is a pretty big PR and reviewing it thoroughly might not be possible, because it would take too long. But I'd appreciate it if you could take a quick look and maybe tell me if anything is still unclear. I can always add more explanations in comments if necessary. 

---
## Things I plan on doing after this PR has been merged
*(I'd create an issue for each of those points)*

- Make this work with in-band lifetimes (there is a TODO comment about this in the code)
- Improve the test suite, in particular:
  - More unit tests
  - More compile_test tests
  - Add compile-fail tests
  - Make `cargo test` execute the compile-test and compile-fail tests
- Create better documentation and update the README, if necessary
- A subset of the proc-macro API will be stabilized for Rust 2018 (see https://github.com/rust-lang/rust/pull/52081). Unfortunately, the API in `proc_macro` that allows the nice kind of diagnostics shown above won't be stabilized by that PR and will thus probably remain unstable for a while. We probably should talk about whether we want this to work on stable Rust 2018 or not. Of course, the best solution would be to use awesome diagnostics on nightly and fall back to a less nice solution on stable. But I'm not yet sure how difficult it would be to use such a fallback mechanism. 
- Possibly more, but I can't think of anything I missed...
